File: android-core/src/main/java/com/mparticle/internal/listeners/InternalListenerManager.java
package com.mparticle.internal.listeners;

import android.content.ContentValues;
import android.content.Context;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.mparticle.SdkListener;
import com.mparticle.identity.AliasResponse;
import com.mparticle.internal.InternalSession;
import com.mparticle.internal.KitFrameworkWrapper;
import com.mparticle.internal.Logger;
import com.mparticle.internal.MPUtility;

import org.json.JSONException;
import org.json.JSONObject;

import java.lang.ref.WeakReference;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class InternalListenerManager implements InternalListener {
    private static InternalListenerManager instance = null;
    private static final String INTERNAL_LISTENER_PROP = "debug.mparticle.listener";
    private Context context;
    final List<WeakReference<SdkListener>> sdkListeners = new ArrayList<WeakReference<SdkListener>>();
    final List<WeakReference<GraphListener>> graphListeners = new ArrayList<WeakReference<GraphListener>>();
    private boolean thrown = false;

    private InternalListenerManager(Context context) {
        this.context = context;
    }

    @Nullable
    public static InternalListenerManager start(Context context) {
        boolean canRun = MPUtility.isAppDebuggable(context) || context.getPackageName().equals(MPUtility.getProp(INTERNAL_LISTENER_PROP));
        if (instance == null && context != null && canRun) {
            instance = new InternalListenerManager(context.getApplicationContext());
        }
        return instance;
    }

    public static boolean isEnabled() {
        return instance != null &&
                instance.hasListeners();
    }

    @NonNull
    public static InternalListener getListener() {
        if (instance != null && isEnabled()) {
            return instance;
        } else {
            return InternalListener.EMPTY;
        }
    }

    public void addListener(SdkListener sdkListener) {
        for (WeakReference<SdkListener> listener : sdkListeners) {
            if (listener.get() == sdkListener) {
                return;
            }
        }
        sdkListeners.add(new WeakReference<SdkListener>(sdkListener));
        if (sdkListener instanceof GraphListener) {
            graphListeners.add(new WeakReference<GraphListener>((GraphListener) sdkListener));
        }
    }

    public void removeListener(SdkListener sdkListener) {
        for (WeakReference<SdkListener> listener : new ArrayList<WeakReference<SdkListener>>(sdkListeners)) {
            if (listener.get() == sdkListener) {
                sdkListeners.remove(listener);
            }
        }
        for (WeakReference<GraphListener> listener : new ArrayList<WeakReference<GraphListener>>(graphListeners)) {
            if (listener.get() == sdkListener) {
                graphListeners.remove(listener);
            }
        }
    }

    @Override
    public void onApiCalled(final Object... objects) {
        try {
            StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
            int index = -1;
            for (int i = 0; i < stackTrace.length; i++) {
                StackTraceElement element = stackTrace[i];
                if (element.getClassName().equals(getClass().getName())) {
                    index = i;
                    break;
                }
            }
            //move to call after the last call in this class;
            index++;
            while (stackTrace[index].getClassName().equals(ListenerHook.class.getName())) {
                index++;
            }
            StackTraceElement apiCall = stackTrace[index];
            StackTraceElement calledFrom = stackTrace[index + 1];
            final boolean isExternalApiInvocation = isExternalApiInvocation(calledFrom);
            final String apiName = getApiName(apiCall);

            broadcast(new SdkListenerRunnable() {
                @Override
                public void run(SdkListener listener) {
                    listener.onApiCalled(apiName, Arrays.asList(objects), isExternalApiInvocation);
                }
            });
        } catch (Exception ex) {
            if (!thrown) {
                thrown = true;
                Logger.error("SdkListener failed onApiCalled!\n" + ex.getMessage());
            }
        }
    }

    @Override
    public void onKitApiCalled(int kitId, Boolean used, Object... objects) {
        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        String methodName = null;
        for (int i = 0; i < stackTrace.length; i++) {
            StackTraceElement element = stackTrace[i];
            if (element.getClassName().equals(KitFrameworkWrapper.class.getName())) {
                methodName = element.getMethodName() + "()";
            }
        }
        onKitApiCalled(stackTrace, methodName, kitId, used, objects);
    }

    @Override
    public void onKitApiCalled(String methodName, int kitId, Boolean used, Object... objects) {
        onKitApiCalled(Thread.currentThread().getStackTrace(), methodName, kitId, used, objects);
    }

    private void onKitApiCalled(StackTraceElement[] stackTrace, final String methodName, final int kitId, final boolean used, Object... objects) {
        String invokingApiMethodName = null;
        String kitManagerMethodName = null;
        boolean foundInternal = false;
        boolean foundExternal = false;
        final List<Object> objectList = new ArrayList<Object>();
        for (Object obj : objects) {
            objectList.add(obj);
        }
        for (int i = 0; i < stackTrace.length; i++) {
            if (!isExternalApiInvocation(stackTrace[i])) {
                foundInternal = true;
            }
            if (foundInternal && !foundExternal) {
                if (isExternalApiInvocation(stackTrace[i])) {
                    invokingApiMethodName = getApiName(stackTrace[i - 1]);
                    foundExternal = true;
                }
            }
            if (stackTrace[i].getClassName().equals("com.mparticle.kits.KitManagerImpl")) {
                kitManagerMethodName = getApiName(stackTrace[i]);
            }
        }
        final String finalInvokingApiMethodName = invokingApiMethodName;
        final String finalKitManagerMethodName = kitManagerMethodName;
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onKitApiCalled(kitId, methodName, finalInvokingApiMethodName, finalKitManagerMethodName, objectList, used);
            }
        });
    }


    @Override
    public void onCompositeObjects(final Object child, final Object parent) {
        broadcast(new SdkGraphListenerRunnable() {
            @Override
            public void run(GraphListener listener) {
                listener.onCompositeObjects(child, parent);
            }
        });
    }

    @Override
    public void onThreadMessage(final String handlerName, final Message msg, final boolean onNewThread) {
        StackTraceElement[] stackTrace = null;
        if (!onNewThread) {
            stackTrace = Thread.currentThread().getStackTrace();
        }
        final StackTraceElement[] finalStackTrace = stackTrace;
        broadcast(new SdkGraphListenerRunnable() {
            @Override
            public void run(GraphListener listener) {
                listener.onThreadMessage(handlerName, msg, onNewThread, finalStackTrace);
            }
        });
    }

    @Override
    public void onEntityStored(final Long primaryKey, final String tableName, ContentValues contentValues) {
        onCompositeObjects(contentValues, tableName + primaryKey);
        final JSONObject jsonObject = new JSONObject();
        SdkListener.DatabaseTable table = null;
        try {
            table = SdkListener.DatabaseTable.valueOf(tableName.toUpperCase());
        } catch (IllegalArgumentException ex) {
            table = SdkListener.DatabaseTable.UNKNOWN;
        }
        for (Map.Entry<String, Object> entry : contentValues.valueSet()) {
            try {
                if (entry.getValue() == null) {
                    jsonObject.put(entry.getKey(), JSONObject.NULL);
                } else {
                    jsonObject.put(entry.getKey(), entry.getValue());
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        final SdkListener.DatabaseTable finalTable = table;
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onEntityStored(finalTable, primaryKey, jsonObject);
            }
        });
    }

    @Override
    public void onNetworkRequestStarted(final SdkListener.Endpoint type, final String url, final JSONObject body, Object... objects) {
        for (Object obj : objects) {
            onCompositeObjects(obj, body);
        }
        final List<Object> objectList = new ArrayList<Object>();
        for (Object obj : objects) {
            objectList.add(obj);
        }
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onNetworkRequestStarted(type, url, body);
            }
        });
    }

    @Override
    public void onNetworkRequestFinished(final SdkListener.Endpoint type, final String url, final JSONObject response, final int responseCode) {
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onNetworkRequestFinished(type, url, response, responseCode);
            }
        });
    }

    @Override
    public void onSessionUpdated(final InternalSession internalSession) {
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onSessionUpdated(new InternalSession(internalSession));
            }
        });
    }

    @Override
    public void onKitDetected(final int kitId) {
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onKitDetected(kitId);
            }
        });
    }

    @Override
    public void onKitConfigReceived(final int kitId, final String configuration) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject = new JSONObject(configuration);
        } catch (JSONException e) {
        }
        final JSONObject jsonConfig = jsonObject;
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onKitConfigReceived(kitId, jsonConfig);
            }
        });
    }

    @Override
    public void onKitExcluded(final int kitId, final String reason) {
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onKitExcluded(kitId, reason);
            }
        });
    }

    @Override
    public void onKitStarted(final int kitId) {
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onKitStarted(kitId);
            }
        });
    }

    @Override
    public void onAliasRequestFinished(final AliasResponse aliasResponse) {
        broadcast(new SdkListenerRunnable() {
            @Override
            public void run(SdkListener listener) {
                listener.onAliasRequestFinished(aliasResponse);
            }
        });
    }

    private void broadcast(SdkListenerRunnable runnable) {
        for (WeakReference<SdkListener> listenerRef : new ArrayList<WeakReference<SdkListener>>(sdkListeners)) {
            SdkListener listener = listenerRef.get();
            if (listener == null) {
                sdkListeners.remove(listenerRef);
            } else {
                runnable.run(listener);
            }
        }
    }

    private void broadcast(SdkGraphListenerRunnable runnable) {
        for (WeakReference<GraphListener> listenerRef : new ArrayList<WeakReference<GraphListener>>(graphListeners)) {
            GraphListener listener = listenerRef.get();
            if (listener == null) {
                graphListeners.remove(listenerRef);
            } else {
                runnable.run(listener);
            }
        }
    }

    interface SdkListenerRunnable {
        void run(SdkListener listener);
    }

    interface SdkGraphListenerRunnable {
        void run(GraphListener listener);
    }


    private String getApiName(StackTraceElement element) {
        String classNameString = getClassName(element.getClassName(), element.getMethodName());
        return getApiFormattedName(classNameString, element.getMethodName());
    }

    public static String getApiFormattedName(String className, String methodName) {
        return new StringBuilder()
                .append(className)
                .append(".")
                .append(methodName)
                .append("()")
                .toString();
    }

    private String getClassName(String className, String methodName) {
        String[] packageNames = className.split("\\.");
        String simpleClassName = packageNames[packageNames.length - 1];
        if (isObfuscated(simpleClassName)) {
            try {
                List<Class> superClasses = new ArrayList<Class>();
                Class clazz = Class.forName(className);
                superClasses.add(clazz.getSuperclass());
                for (Class interfce : clazz.getInterfaces()) {
                    superClasses.add(interfce);
                }
                for (Class superClass : superClasses) {
                    for (Method method : superClass.getMethods()) {
                        if (method.getName().equals(methodName)) {
                            String superClassName = getClassName(superClass.getName(), methodName);
                            if (!isObfuscated(superClassName)) {
                                return superClassName;
                            }
                        }
                    }
                }
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
        }
        return simpleClassName;
    }

    private boolean isObfuscated(@NonNull String className) {
        return Character.isLowerCase(className.toCharArray()[0]) && className.length() <= 3;
    }

    private boolean isExternalApiInvocation(StackTraceElement element) {
        return !element.getClassName().startsWith("com.mparticle") ||
                (element.getClassName().startsWith(context.getApplicationContext().getPackageName()) &&
                        context.getApplicationContext().getPackageName().length() > 1);
    }

    private boolean hasListeners() {
        return instance.sdkListeners.size() > 0 || instance.graphListeners.size() > 0;
    }
}

File: android-core/src/main/kotlin/com/mparticle/internal/listeners/InternalListenerManager.kt
package com.mparticle.internal.listeners

import android.content.ContentValues
import android.content.Context
import android.os.Message
import com.mparticle.SdkListener
import com.mparticle.identity.AliasResponse
import com.mparticle.internal.InternalSession
import com.mparticle.internal.KitFrameworkWrapper
import com.mparticle.internal.MPUtility
import org.json.JSONException
import org.json.JSONObject
import java.lang.ref.WeakReference
import java.util.Locale

class InternalListenerManager private constructor(private val context: Context) : InternalListener {
    val sdkListeners: MutableList<WeakReference<SdkListener>> = ArrayList()
    val graphListeners: MutableList<WeakReference<GraphListener>> = ArrayList()
    private val thrown = false

    fun addListener(sdkListener: SdkListener) {
        for (listener in sdkListeners) {
            if (listener.get() === sdkListener) {
                return
            }
        }
        sdkListeners.add(WeakReference(sdkListener))
        if (sdkListener is GraphListener) {
            graphListeners.add(WeakReference(sdkListener as GraphListener))
        }
    }

    fun removeListener(sdkListener: SdkListener) {
        for (listener in ArrayList(sdkListeners)) {
            if (listener.get() === sdkListener) {
                sdkListeners.remove(listener)
            }
        }
        for (listener in ArrayList(graphListeners)) {
            if (listener.get() === sdkListener) {
                graphListeners.remove(listener)
            }
        }
    }

    override fun onApiCalled(vararg objects: Any) {
    }

    override fun onKitApiCalled(kitId: Int, used: Boolean, vararg objects: Any) {
        val stackTrace = Thread.currentThread().stackTrace
        var methodName: String? = null
        for (i in stackTrace.indices) {
            val element = stackTrace[i]
            if (element.className == KitFrameworkWrapper::class.java.name) {
                methodName = element.methodName + "()"
            }
        }
        methodName?.let { onKitApiCalled(stackTrace, it, kitId, used, *objects) }
    }

    override fun onKitApiCalled(
        methodName: String, kitId: Int, used: Boolean, vararg objects: Any
    ) {
        onKitApiCalled(Thread.currentThread().stackTrace, methodName, kitId, used, *objects)
    }

    private fun onKitApiCalled(
        stackTrace: Array<StackTraceElement>,
        methodName: String,
        kitId: Int,
        used: Boolean,
        vararg objects: Any
    ) {
        var invokingApiMethodName: String? = null
        var kitManagerMethodName: String? = null
        var foundInternal = false
        var foundExternal = false
        val objectList: MutableList<Any> = ArrayList()
        for (obj in objects) {
            objectList.add(obj)
        }
        for (i in stackTrace.indices) {
            if (!isExternalApiInvocation(stackTrace[i])) {
                foundInternal = true
            }
            if (foundInternal && !foundExternal) {
                if (isExternalApiInvocation(stackTrace[i])) {
                    invokingApiMethodName = getApiName(stackTrace[i - 1])
                    foundExternal = true
                }
            }
            if (stackTrace[i].className == "com.mparticle.kits.KitManagerImpl") {
                kitManagerMethodName = getApiName(stackTrace[i])
            }
        }
        val finalInvokingApiMethodName = invokingApiMethodName
        val finalKitManagerMethodName = kitManagerMethodName
        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                listener.onKitApiCalled(
                    kitId,
                    methodName,
                    finalInvokingApiMethodName,
                    finalKitManagerMethodName,
                    objectList,
                    used
                )
            }
        })
    }

    override fun onCompositeObjects(child: Any?, parent: Any?) {
        broadcast(object : SdkGraphListenerRunnable {
            override fun run(listener: GraphListener) {
                listener.onCompositeObjects(child, parent)
            }
        })
    }

    override fun onThreadMessage(handlerName: String, msg: Message, onNewThread: Boolean) {
        var stackTrace: Array<StackTraceElement?>? = null
        if (!onNewThread) {
            stackTrace = Thread.currentThread().stackTrace
        }
        val finalStackTrace = stackTrace
        broadcast(object : SdkGraphListenerRunnable {
            override fun run(listener: GraphListener) {
                listener.onThreadMessage(handlerName, msg, onNewThread, finalStackTrace)
            }
        })
    }

    override fun onEntityStored(
        primaryKey: Long, tableName: String, contentValues: ContentValues?
    ) {
        onCompositeObjects(contentValues, tableName + primaryKey)
        val jsonObject = JSONObject()
        var table: SdkListener.DatabaseTable? = null
        table = try {
            tableName.uppercase(Locale.getDefault()).let { SdkListener.DatabaseTable.valueOf(it) }
        } catch (ex: IllegalArgumentException) {
            SdkListener.DatabaseTable.UNKNOWN
        }
        if (contentValues != null) {
            for ((key, value) in contentValues.valueSet()) {
                try {
                    if (value == null) {
                        jsonObject.put(key, JSONObject.NULL)
                    } else {
                        jsonObject.put(key, value)
                    }
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }
        }
        val finalTable = table
        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                finalTable?.let { listener.onEntityStored(it, primaryKey, jsonObject) }
            }
        })
    }

    override fun onNetworkRequestStarted(
        type: SdkListener.Endpoint, url: String, body: JSONObject?, vararg objects: Any
    ) {
        for (obj in objects) {
            onCompositeObjects(obj, body)
        }
        val objectList: MutableList<Any> = ArrayList()
        for (obj in objects) {
            objectList.add(obj)
        }
        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                listener.onNetworkRequestStarted(type, url, body ?: JSONObject())
            }
        })
    }

    override fun onNetworkRequestFinished(
        type: SdkListener.Endpoint, url: String, response: JSONObject?, responseCode: Int
    ) {

        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                listener.onNetworkRequestFinished(type, url, response, responseCode)
            }
        })
    }

    override fun onSessionUpdated(internalSession: InternalSession) {
        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                listener.onSessionUpdated(InternalSession(internalSession))
            }
        })
    }

    override fun onKitDetected(kitId: Int) {
        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                listener.onKitDetected(kitId)
            }
        })
    }

    override fun onKitConfigReceived(kitId: Int, configuration: String?) {
        var jsonObject = JSONObject()
        try {
            jsonObject = JSONObject(configuration)
        } catch (e: JSONException) {
        }
        val jsonConfig = jsonObject
        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                listener.onKitConfigReceived(kitId, jsonConfig)
            }
        })
    }

    override fun onKitExcluded(kitId: Int, reason: String?) {
        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                listener.onKitExcluded(kitId, reason)
            }
        })
    }

    override fun onKitStarted(kitId: Int) {
        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                listener.onKitStarted(kitId)
            }
        })
    }

    override fun onAliasRequestFinished(aliasResponse: AliasResponse?) {
        broadcast(object : SdkListenerRunnable {
            override fun run(listener: SdkListener) {
                listener.onAliasRequestFinished(aliasResponse)
            }
        })
    }

    private fun broadcast(runnable: SdkListenerRunnable) {
        for (listenerRef in ArrayList(sdkListeners)) {
            val listener = listenerRef.get()
            if (listener == null) {
                sdkListeners.remove(listenerRef)
            } else {
                runnable.run(listener)
            }
        }
    }

    private fun broadcast(runnable: SdkGraphListenerRunnable) {
        for (listenerRef in ArrayList(graphListeners)) {
            val listener = listenerRef.get()
            if (listener == null) {
                graphListeners.remove(listenerRef)
            } else {
                runnable.run(listener)
            }
        }
    }

    internal interface SdkListenerRunnable {
        fun run(listener: SdkListener)
    }

    internal interface SdkGraphListenerRunnable {
        fun run(listener: GraphListener)
    }

    private fun getApiName(element: StackTraceElement): String {
        val classNameString = getClassName(element.className, element.methodName)
        return getApiFormattedName(classNameString, element.methodName)
    }

    private fun getClassName(className: String, methodName: String): String {
        val packageNames =
            className.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val simpleClassName = packageNames[packageNames.size - 1]
        if (isObfuscated(simpleClassName)) {
            try {
                val superClasses: MutableList<Class<*>> = ArrayList()
                val clazz = Class.forName(className) // nosemgrep
                superClasses.add(clazz.superclass)
                for (interfce in clazz.interfaces) {
                    superClasses.add(interfce)
                }
                for (superClass in superClasses) {
                    for (method in superClass.methods) {
                        if (method.name == methodName) {
                            val superClassName = getClassName(superClass.name, methodName)
                            if (!isObfuscated(superClassName)) {
                                return superClassName
                            }
                        }
                    }
                }
            } catch (e: ClassNotFoundException) {
                e.printStackTrace()
            }
        }
        return simpleClassName
    }

    private fun isObfuscated(className: String): Boolean {
        return Character.isLowerCase(className.toCharArray()[0]) && className.length <= 3
    }

    private fun isExternalApiInvocation(element: StackTraceElement): Boolean {
        return !element.className.startsWith("com.mparticle") || (element.className.startsWith(
            context.applicationContext.packageName
        ) && context.applicationContext.packageName.length > 1)
    }

    private fun hasListeners(): Boolean {
        return (instance?.sdkListeners?.size ?: 0) > 0 || (instance?.graphListeners?.size ?: 0) > 0
    }

    companion object {
        private var instance: InternalListenerManager? = null
        private const val INTERNAL_LISTENER_PROP = "debug.mparticle.listener"

        @JvmStatic
        fun start(context: Context?): InternalListenerManager? {
            val canRun = MPUtility.isAppDebuggable(context) || context?.packageName == MPUtility.getProp(INTERNAL_LISTENER_PROP)
            if (instance == null && context != null && canRun) {
                instance = InternalListenerManager(context.applicationContext)
            }
            return instance
        }

        @JvmStatic
        val isEnabled: Boolean
            get() = instance != null && instance?.hasListeners() == true

        @JvmStatic
        val listener: InternalListener
            get() = instance?.let {
                if (isEnabled) it else InternalListener.EMPTY
            } ?: InternalListener.EMPTY

        fun getApiFormattedName(className: String?, methodName: String?): String {
            return StringBuilder().append(className).append(".").append(methodName).append("()")
                .toString()
        }
    }
}
